usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -1107,16 +1117,19 @@ zil_lwb_commit(zilog_t *zilog, itx_t *itx, lwb_t *lwb)
 			char *dbuf;
 			int error;
 
-			if (dlen) {
-				ASSERT(itx->itx_wr_state == WR_NEED_COPY);
+			if (itx->itx_wr_state == WR_NEED_COPY) {
 				dbuf = lr_buf + reclen;
-				lrw->lr_common.lrc_reclen += dlen;
+				lrcb->lrc_reclen += dnow;
+				if (lrwb->lr_length > dnow)
+					lrwb->lr_length = dnow;
+				lrw->lr_offset += dnow;
+				lrw->lr_length -= dnow;
```

Comment:
May be, but I don't want to change the logic of the code much at this point without significant reason.  This code is already committed to FreeBSD and diverging from it makes testing difficult for me.  Lets do it this "good" way which was well tested, remove divergence and then look for "better".